pkg/storage/mvcc.go

```diff
@@ -5332,6 +5332,144 @@ func CanGCEntireRange(
 	return coveredByRangeTombstones, nil
 }
 
+var emptyMetaToken enginepb.MVCCMetadata
+
+// MVCCGarbageCollectPointsWithClearRange removes garbage collected data falling
+// under ranges covered by [start@startTimestamp, endTimestamp). This function
+// performs a check to ensure that no non-deleted data (most recent or history
+// with timestamp greater that threshold) is being deleted.
+func MVCCGarbageCollectPointsWithClearRange(
+	ctx context.Context,
+	rw ReadWriter,
+	ms *enginepb.MVCCStats,
+	start, end roachpb.Key,
+	startTimestamp hlc.Timestamp,
+	gcThreshold hlc.Timestamp,
+) error {
+	var countKeys int64
+	var removedEntries int64
+	defer func(begin time.Time) {
+		// TODO(oleg): this could be misleading if GC fails, but this function still
+		// reports how many keys were GC'd. The approach is identical to what point
+		// key GC does for consistency, but both places could be improved.
+		log.Eventf(ctx,
+			"done with GC evaluation for clear range of %d keys at %.2f keys/sec. Deleted %d entries",
+			countKeys, float64(countKeys)*1e9/float64(timeutil.Since(begin)), removedEntries)
+	}(timeutil.Now())
+
+	iter := rw.NewMVCCIterator(MVCCKeyAndIntentsIterKind, IterOptions{
+		LowerBound: start,
+		UpperBound: end,
+		KeyTypes:   IterKeyTypePointsAndRanges,
+	})
+	defer iter.Close()
+
+	iter.SeekGE(MVCCKey{Key: start})
+
+	var (
+		prevPointKey    MVCCKey
```

Comment:
It would be good to explain how this works, I had to read the logic a few times to understand it. In particular, when encountering a new key this no longer points to the previous key, but the current key.